discoverd/client/watch.go

```diff
@@ -0,0 +1,175 @@
+package discoverd
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/flynn/flynn/pkg/attempt"
+	"github.com/flynn/flynn/pkg/stream"
+)
+
+type WatchState string
+
+const (
+	WatchStateConnected    WatchState = "connected"
+	WatchStateDisconnected            = "disconnected"
+)
+
+// Watch is a wrapper around an event stream which reconnects on error and
+// generates any events which were missed by comparing local state to the
+// current state returned by the server.
+type Watch struct {
+	instances   map[string]*Instance
+	serviceMeta *ServiceMeta
+	stream      stream.Stream
+	err         error
+	done        chan struct{}
+	doneOnce    sync.Once
+	stateCh     chan WatchState
+}
+
+func NewWatch() *Watch {
+	return &Watch{
+		instances: make(map[string]*Instance),
+		done:      make(chan struct{}),
+	}
+}
+
+// SetStateChannel is used in tests to set a channel which will receive events
+// when the watch connects to / disconects from the server.
+func (w *Watch) SetStateChannel(ch chan WatchState) {
+	w.stateCh = ch
```

Comment:
This isn't thread-safe. The comment should note it (making it only useful for tests) or it should use a `atomic.Value`.